### PR TITLE
perplexity metric callback

### PR DESCRIPTION
**Pull request by elephantmipt**
## Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md), Pull Request section?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [x] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?

## Description

SImple and useful perplexity metric callback

## Related Issue

https://github.com/catalyst-team/catalyst/issues/805
## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

**You can use 'Login as guest' to see Teamcity build logs.**

### CODE REVIEW

File: catalyst/contrib/dl/callbacks/perplexity.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+from torch import nn
+
+from catalyst.core.callbacks import MetricCallback
+
+
+class PerplexityCallback(MetricCallback):
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/perplexity.py]: what do you this about `PerplexityMetricCallback`?

File: catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+from catalyst import dl
+from catalyst.contrib.data.nlp import LanguageModelingDataset
+from catalyst.contrib.dl.callbacks import PerplexityCallback
+
+
+class HuggingFaceRunner(dl.Runner):
+    """Just an example"""
+
+    def _handle_batch(self, batch):
+        mlm_labels = batch.get("masked_lm_labels")
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py]: maybe, let's use full name?

**Review by Scitator — APPROVED** [review]
could you please update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?

File: catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import torch
+from transformers import AutoModelWithLMHead, AutoTokenizer
+from transformers.data.data_collator import DataCollatorForLanguageModeling
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py:3]: due to `transformers`  dependency, you need to add this test to https://github.com/catalyst-team/catalyst/blob/master/bin/tests/check_dl_nlp.sh rather then typical tests.
we could not run them correctly with core Catalyst version.

**Review by Scitator — CHANGES_REQUESTED** [review]

File: catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py (hunk 3 above)
Comment (Scitator) [reply catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py:3]: :thinking: deal, rechecked :)
need to think about it a bit more, test looks valid in the end

Comment (elephantmipt) [reply catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py:3]: `transformers` installed due to this line:
https://github.com/catalyst-team/catalyst/blob/master/.github/workflows/codestyle.yml#L56
as it is in requirements_nlp

Comment (elephantmipt) [reply catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py:3]: Only codestyle test runs pytest and I need transformers only for tests. I see no problem with dependencies here because callback will also work correct without transformers library. What about to merge it?)

File: catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+def test_is_running():
+    """Test if perplexity is running normal"""
+    tok = AutoTokenizer.from_pretrained("distilbert-base-uncased")
+    model = AutoModelWithLMHead.from_pretrained("distilbert-base-uncased")
+    dataset = LanguageModelingDataset(texts, tok)
+    collate_fn = DataCollatorForLanguageModeling(tok).collate_batch
+    dataloader = torch.utils.data.DataLoader(dataset, collate_fn=collate_fn)
+    optimizer = torch.optim.Adam(model.parameters(), lr=5e-5)
+
+    runner = HuggingFaceRunner()
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py:48]: could you please add `check=True` to make test training faster?

**Review by Scitator — APPROVED** [review]

File: catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py [hunk 5]
```diff
@@ -45,7 +45,7 @@ def test_is_running():
     dataloader = torch.utils.data.DataLoader(dataset, collate_fn=collate_fn)
     optimizer = torch.optim.Adam(model.parameters(), lr=5e-5)
 
-    runner = HuggingFaceRunner()
+    runner = HuggingFaceRunner(check=True)
     runner.train(
```
Comment (Scitator) [inline catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py:49]: you need to pass `check=True` to `.train`

Comment (elephantmipt) [reply catalyst/contrib/dl/callbacks/tests/test_perplexity_callback.py:49]: yep, already fixed it. Just misclicked)
